### PR TITLE
[hail] Experimental support for compiling comparators

### PR DESCRIPTION
**Pull request by danking**
Dice came up @patrick-schultz.

- Parameterize (de/en)coders by InputStream type which is either a Java input stream or an array of bytes.
- add `compileComparison` which produces the bytes of a system-specific executable for comparing two values of types `l` and `r` encoded in `codec`
- experimental python API to `compileComparison` to enable experimentation

The end goal is to run `compileComparison` and ship the bytes to the shuffle service which will use it as a comparison operation on the encoded keys.

### CODE REVIEW

File: hail/src/test/scala/is/hail/nativecode/NativeDecoderSuite.scala [hunk 1]
```diff
@@ -82,7 +82,7 @@ class NativeDecoderSuite extends HailSuite {
            |auto jis = h->at(0);
            |
            |auto is = std::make_shared<InputStream>(up, jis);
-           |${ spec.nativeInputBufferType } leb_buf {is};
+           |${ spec.nativeInputBufferType _ } leb_buf {is};
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/nativecode/NativeDecoderSuite.scala]: I'm confused. Shouldn't this be `spec.nativeInputBufferType("InputStream")`?

Comment (danking) [reply hail/src/test/scala/is/hail/nativecode/NativeDecoderSuite.scala]: yep. bad rebase.

File: hail/src/main/resources/include/hail/Decoder.h [hunk 2]
```diff
@@ -91,22 +108,36 @@ class StreamInputBuffer {
     bool read_boolean() { return read_byte() != 0; };
 };
 
-template <int BUFSIZE, typename InputBlockBuffer>
+template <typename IS>
+StreamInputBlockBuffer<IS>::StreamInputBlockBuffer(std::shared_ptr<IS> is) :
+  input_stream_(is) { }
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/src/main/resources/include/hail/Decoder.h]: Will do. I just copy pasted it in from the cpp file when I added a template

File: hail/src/main/c/Decoder.cpp [hunk 3]
```diff
@@ -49,17 +50,24 @@ InputStream::~InputStream() {
   up_.env()->DeleteGlobalRef(jinput_stream_);
 }
 
-// StreamInputBlockBuffer
-StreamInputBlockBuffer::StreamInputBlockBuffer(std::shared_ptr<InputStream> is) :
-  input_stream_(is) { }
+ByteArrayInputStream::ByteArrayInputStream(char *bytes, long size) :
+  bytes{bytes}, cursor{0}, size{size} {}
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/src/main/c/Decoder.cpp]: I agree with you, but this follows the style I observe here: non-template things go out-of-line in .cpp, template things in-line in .h.

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/c/Decoder.cpp]: We don't follow that style strictly (nor should we). There are lots of inline constructors and methods even in `Decoder.h`. Rule-of-thumb should be: methods you want the compiler to have the option of inlining must go in the header file, and weigh in-class vs. out-of-class by readability. On the other hand, too many, and too large, methods in the header increases compilation times.

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/c/Decoder.cpp]: Oh sorry, you're right that in `Decoder.h` all the inline things are templated.

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/c/Decoder.cpp]: I take it back: see `StreamInputBuffer`

[the remark this answers is not in the document]
Comment (danking) [reply hail/src/main/c/Decoder.cpp]: all moved

**Pull request merged**